File: 206-reverse-linked-list/reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reverseList(self, head):
        if not head:
            return None
        elif not head.next:
            return head
        
        tmp_l=None
        tmp_r=tmp=head
        while tmp.next:
            tmp_r=tmp.next
            tmp.next=tmp_l
            tmp_l=tmp
            tmp = tmp_r
        tmp.next = tmp_l
        return tmp







        

File: 206-reverse-linked-list/reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        if(head==null){
            return null;
        }
        else if(head.next==null){
            return head;
        }
        else{
            ListNode headNode = recurReverse(head, head.next);
            head.next = null;
            return headNode;
        }
    }

    public ListNode recurReverse(ListNode head, ListNode next){
        if(next.next==null){
            head.next=null;
            next.next=head;
            return next;
        }
        ListNode headNode = recurReverse(head.next, next.next);
        head.next=null;
        next.next=head;
        return headNode;
    }
}